### PR TITLE
Merge Shapelets and Zernike Polynomials into Master

### PR DESCRIPTION
**Pull request by JoshVStaden**
- [x] Tests added / passed

  ```bash
  $ py.test -v -s africanus
  ```

  If the pep8 tests fail, the quickest way to correct
  this is to run `autopep8` and then `flake8` and
  `pycodestyle` to fix the remaining issues.

  ```
  $ pip install -U autopep8 flake8 pycodestyle
  $ autopep8 -r -i africanus
  $ flake8 africanus
  $ pycodestyle africanus
  ```

- [ ] Fully documented, including `HISTORY.rst` for all changes
      and one of the `docs/*-api.rst` files for new API

  To build the docs locally:

  ```
  pip install -r requirements.readthedocs.txt
  cd docs
  READTHEDOCS=True make html
  ```

### CODE REVIEW

**Comment by JoshVStaden** [comment]
@sjperkins Hi Simon. I have created the PR for this branch. I have removed the dependency of that shapelets package and lowered the amount of test data in the shapelets test. It also seemed that using the SciPy fft in the test cases caused some issues, so I changed them to use the NumPy fft, and it worked without problems. It seems that particular test file takes about 2 seconds to run, and the other test files take about a minute to run. Is this an okay amount, or should this be lowered? 

All that is left now is to just host the binary data for the Zernike beam onto PackRATT. I want to carry on with making the benchmark for the thesis, and then I will get onto that. On that note, I am having issues with specifying the number of CPU cores for Eidos to run on. As far as I can tell, there is no command included in Eidos for this. It seems that Eidos just uses the usual NumPy functions to do multithreading. I have tried searching for a way to force NumPy to use a specific number of cores, but I cannot find this anywhere. Is there something that I could use to do this?

**Comment by landmanbester** [comment]
Weird. I don't see the option but I remember seeing some multi-processing code in there at some point. Maybe it tries to use all available resources. Can you check if multiple threads spin up when you run it?

**Comment by landmanbester** [comment]
Unless [this](https://github.com/ratt-ru/eidos/blob/master/eidos/parallelize.py) is only used for the decomposition (not reconstruction)

**Comment by JoshVStaden** [comment]
It seems that the parallelization does get called for the beam reconstruction. For one frequency, Eidos only runs on one core. However, specifying more than one frequency uses the parallelizing code. I assume then that Eidos parallelizes on frequency. I can also see how to specify the number of CPU cores for Eidos to use. So I think that I should generate a beam for 32 frequencies for each implementation and just vary the number of cores used.

**Comment by sjperkins** [comment]
> @sjperkins Hi Simon. I have created the PR for this branch. I have removed the dependency of that shapelets package and lowered the amount of test data in the shapelets test. It also seemed that using the SciPy fft in the test cases caused some issues, so I changed them to use the NumPy fft, and it worked without problems. It seems that particular test file takes about 2 seconds to run, and the other test files take about a minute to run. Is this an okay amount, or should this be lowered?

@JoshVStaden Are you saying the final reduced test is 2 seconds (this is great) compared to the original run of about a minute? If so that is fine. If there's a test that still takes a minute, it will need to be further reduced.
 
> All that is left now is to just host the binary data for the Zernike beam onto PackRATT.

So this can simply be a binary .tar.gz. Could you please 

1. Host this on google drive or elwood's FTP
2. Define a registry entry for packratt (as specified [here](https://github.com/ratt-ru/packratt#registry-schemas).
3. Place the registry entry in ``$HOME/.config/packratt/registry.yaml`` and test that it works with your test case.
4. Modify your test case to use packratt to get the file and use the contents?

Once that works, we can host the test data in a more permanent location.

**Comment by JoshVStaden** [comment]
@sjperkins Hi Simon. I have passed the flake8 tests. It seems that the CI server is only reporting a single error now, This seems to be coming from africanus/rime/tests/test_wsclean_predict.py::test_dask_wsclean_predict. Apparently, when calling dask_wsclean_predict, the test script uses the wrong number of parameters (9 instead of 7). I tried a quick fix for this on my local files to no avail. Once it ran the code, a lot of Numba errors appeared due to unsupported operations. 

I don't want to fiddle too much on code that I haven't created myself, so I was wondering if this is something that can be handled your side?

**Comment by sjperkins** [comment]
Hi @JoshVStaden https://github.com/ska-sa/codex-africanus/pull/231/commits/d7283904d36b40c24923a0cdec391c388b5cf69c appears to have applied black formatting over the entire code base. I believe, from the Tuesday meeting, that the idea was to apply black formatting to the zernike polynomial code only. Could you please revert and fix?

**Comment by JoshVStaden** [comment]
@sjperkins I have reverted the changes and applied black formatting on my own code. The issue with dask_wsclean_predict seems to be persisting.

**Comment by sjperkins** [comment]
@JoshVStaden I'm not seeing the revert of https://github.com/ska-sa/codex-africanus/commit/d7283904d36b40c24923a0cdec391c388b5cf69c. For instance, if one examines the "Files Changed" tab, the averaging code contains formatting changes. Could you please check again?

**Comment by JoshVStaden** [comment]
@sjperkins Hi Simon. It seems that the code wasn't correctly reverted before. I have reverted it now. The tests are also passing.

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,7 @@ __pycache__/
 # C extensions
 *.so
 
+
 # Distribution / packaging
```
Comment (sjperkins) [inline .gitignore:13]: Revert this whitespace

File: africanus/__init__.py [hunk 2]
```diff
@@ -11,4 +11,4 @@
 
 __author__ = """Simon Perkins"""
 __email__ = 'sperkins@ska.ac.za'
-__version__ = '0.2.9'
+__version__ = '0.2.8'
```
Comment (sjperkins) [inline africanus/__init__.py:14]: Revert this version change

File: africanus/averaging/tests/test_mapping.py [hunk 3]
```diff
@@ -12,7 +12,9 @@
 
 @pytest.fixture
 def time():
-    return np.asarray([1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0])  # noqa
+    return np.asarray(
+        [1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0]
+    )  # noqa
```
Comment (sjperkins) [inline africanus/averaging/tests/test_mapping.py:17]: Revert this formatting change

File: africanus/averaging/tests/test_support.py [hunk 4]
```diff
@@ -10,7 +10,9 @@
 
 @pytest.fixture
 def time():
-    return np.asarray([1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0])  # noqa
+    return np.asarray(
+        [1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0]
+    )  # noqa
```
Comment (sjperkins) [inline africanus/averaging/tests/test_support.py:15]: Revert this formatting change

File: africanus/averaging/tests/test_time_and_channel_averaging.py [hunk 5]
```diff
@@ -16,7 +16,9 @@
 
 @pytest.fixture
 def time():
-    return np.asarray([1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0])  # noqa
+    return np.asarray(
+        [1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0]
+    )  # noqa
```
Comment (sjperkins) [inline africanus/averaging/tests/test_time_and_channel_averaging.py:21]: Revert this formatting change

File: africanus/gridding/wgridder/dask.py [hunk 6]
```diff
@@ -11,13 +11,10 @@
 from africanus.gridding.wgridder.vis2im import DIRTY_DOCS
 from africanus.gridding.wgridder.im2vis import MODEL_DOCS
 from africanus.gridding.wgridder.im2residim import RESIDUAL_DOCS
-from africanus.gridding.wgridder.hessian import HESSIAN_DOCS
```
Comment (sjperkins) [inline africanus/gridding/wgridder/dask.py:14]: Revert removal of hessian code throughout this file

File: africanus/gridding/wgridder/im2residim.py [hunk 7]
```diff
@@ -15,7 +15,7 @@
 @requires_optional('ducc0.wgridder', ducc_import_error)
 def _residual_internal(uvw, freq, image, vis, freq_bin_idx, freq_bin_counts,
                        cell, weights, flag, celly, epsilon, nthreads,
-                       do_wstacking, double_accum):
```
Comment (sjperkins) [inline africanus/gridding/wgridder/im2residim.py:18]: Revert removal of hessian code throughout this file

File: africanus/gridding/wgridder/tests/test_wgridder.py [hunk 8]
```diff
@@ -206,72 +206,6 @@ def test_residual(nx, ny, fov, nrow, nchan, nband,
         residim1/rmax, residim2/rmax, decimal=decimal)
 
 
-@pmp("nx", (128, ))
```
Comment (sjperkins) [inline africanus/gridding/wgridder/tests/test_wgridder.py:209]: Revert removal of hessian code throughout this file

File: africanus/gridding/wgridder/vis2im.py [hunk 9]
```diff
@@ -15,7 +15,7 @@
 @requires_optional('ducc0.wgridder', ducc_import_error)
 def _dirty_internal(uvw, freq, vis, freq_bin_idx, freq_bin_counts, nx, ny,
                     cell, weights, flag, celly, epsilon, nthreads,
-                    do_wstacking, double_accum):
```
Comment (sjperkins) [inline africanus/gridding/wgridder/vis2im.py:18]: Revert removal of hessian code throughout this file

File: africanus/model/spi/dask.py [hunk 10]
```diff
@@ -16,21 +16,20 @@
 
 
 def _fit_spi_components_wrapper(data, weights, freqs, freq0,
```
Comment (sjperkins) [inline africanus/model/spi/dask.py:18]: Revert removal of SPI beam code throughout this file

File: africanus/model/spi/tests/test_component_spi.py [hunk 11]
```diff
@@ -43,12 +43,9 @@ def spi_func(nu, I0, alpha, beam=1.0):
     alphavar2 = np.zeros(ncomps)
 
     for i in range(ncomps):
-        def fit_func(nu, I0, alpha): return spi_func(nu, I0, alpha,
```
Comment (sjperkins) [inline africanus/model/spi/tests/test_component_spi.py:46]: Revert removal of beam SPI code throughout this file

File: africanus/rime/cuda/test_shuffle.py [hunk 12]
```diff
@@ -246,9 +246,11 @@ def test_cuda_shuffle_transpose_2(ncorrs):
             // Horizontal (inter-thread) Rotation
             int addr = case_id;
             {%- for case in range(var_length) %}
-            {{var_name}}[{{case}}] = __shfl_sync(mask, {{var_name}}[{{case}}], addr, {{var_length}});
```
Comment (sjperkins) [inline africanus/rime/cuda/test_shuffle.py:249]: Revert whitespace changes throughout this file

File: africanus/rime/dask.py [hunk 13]
```diff
@@ -1,14 +1,18 @@
 # -*- coding: utf-8 -*-
 
 
-from africanus.rime.phase import (phase_delay as np_phase_delay,
```
Comment (sjperkins) [inline africanus/rime/dask.py:4]: Revert formatting changes throughout this file

File: africanus/rime/examples/predict.py [hunk 14]
```diff
@@ -114,6 +113,7 @@ def baseline_jones_multiply(corrs, *args):
             # Extract it and the next corr index
             einsum_schema, corr_index = schema_fn(corrs, corr_index)
             input_einsum_schemas.append(einsum_schema)
+            print(name, einsum_schema)
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:116]: Remove print statement

File: africanus/rime/examples/predict.py [hunk 15]
```diff
@@ -133,16 +133,14 @@ def create_parser():
     p.add_argument("-rc", "--row-chunks", type=int, default=10000)
     p.add_argument("-mc", "--model-chunks", type=int, default=10)
     p.add_argument("-b", "--beam", default=None)
-    p.add_argument("-l", "--l-axis", default="L")
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:136]: Revert removal of code in this file. (Addition of l and m axis)

File: africanus/rime/examples/predict.py [hunk 16]
```diff
@@ -195,9 +191,7 @@ def __init__(self, filename):
         raise ValueError("FITS must have exactly three axes. "
                          "L or X, M or Y and FREQ. NAXIS != 3")
 
-    (l_ax, l_grid), (m_ax, m_grid), (nu_ax, nu_grid) = beam_grids(header,
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:198]: Revert removal of l_axis and m_axis code in this file

File: africanus/rime/examples/predict.py [hunk 17]
```diff
@@ -284,10 +276,10 @@ def parse_sky_model(filename, chunks):
         try:
             # Extract SPI for I.
             # Zero Q, U and V to get 1 on the exponential
-            spi = [[spectrum.spi]*4]
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:287]: Revert

File: africanus/rime/examples/predict.py [hunk 18]
```diff
@@ -284,10 +276,10 @@ def parse_sky_model(filename, chunks):
...
+            spi = [[spectrum.spi, 0, 0, 0]]
         except AttributeError:
             # Default I SPI to -0.7
-            spi = [[0, 0, 0, 0]]
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:290]: Revert

File: africanus/rime/examples/predict.py [hunk 19]
```diff
@@ -429,8 +420,7 @@ def dde_factory(args, ms, ant, field, pol, lm, utime, frequency):
                        dtype=dtype)
 
     # Load the beam information
-    beam, lm_ext, freq_map = load_beams(args.beam, corr_type,
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:432]: Revert

File: africanus/rime/examples/predict.py [hunk 20]
```diff
@@ -459,15 +449,16 @@ def vis_factory(args, source_type, sky_model,
     uvw = -ms.UVW.data if args.invert_uvw else ms.UVW.data
 
     # (source, row, frequency)
-    phase = phase_delay(lm, uvw, frequency)
+    phase = phase_delay(lm*np.array([[0, 0]]) + np.array([[0, 0.000]]),
```
Comment (sjperkins) [inline africanus/rime/examples/predict.py:452]: This looks incorrect as it zeros all the lm coordinates... Revert

File: setup.py [hunk 21]
```diff
@@ -31,7 +31,7 @@
     'scipy': ['scipy >= 1.4.0'],
     'astropy': ['astropy >= 3.0'],
     'python-casacore': ['python-casacore >= 3.3.1'],
-    'ducc0': ['ducc0 >= 0.8.0'],
+    'ducc0': ['ducc0 >= 0.6.0'],
```
Comment (sjperkins) [inline setup.py:34]: Revert reversion of ducc0 from 0.8.0 to 0.6.0

File: setup.py [hunk 22]
```diff
@@ -88,6 +88,6 @@
     test_suite='tests',
     tests_require=test_requirements,
     url='https://github.com/ska-sa/codex-africanus',
-    version='0.2.9',
+    version='0.2.8',
```
Comment (sjperkins) [inline setup.py:91]: Revert reversion of codex-africanus version from 0.2.9 to 0.2.8

**Review by sjperkins — CHANGES_REQUESTED** [review]
There's still a lot of unnecessary changes @JoshVStaden. Most importantly, this PR currently reverts changes that were recently introduced into master. These must be re-reverted.

This PR  should only contain additions/changes to the zernike/shapelets functionaliy.

**Comment by JoshVStaden** [comment]
@sjperkins I have done the necessary changes.

File: setup.py (hunk 22 above)
Comment (sjperkins) [inline setup.py:91]: Apologies this was a mistake on my part the correct change is 

```suggestion
    version='0.2.9',
```

File: setup.py (hunk 21 above)
Comment (sjperkins) [reply setup.py:34]: I think the double-negative was  confusing here:

```suggestion
    'ducc0': ['ducc0 >= 0.8.0'],
```

**Review by sjperkins — CHANGES_REQUESTED** [review]
Looks good, except for 3 version corrections: I think there was some confusion around the phrasing of reverting the reversions.

**Comment by JoshVStaden** [comment]
@sjperkins I changed the version numbers back

**Review by sjperkins — APPROVED** [review]

**Comment by o-smirnov** [comment]
Yay!

**Pull request merged**